packages/kilt-did/src/types.ts

```diff
@@ -0,0 +1,104 @@
+import type { IIdentity, SubmittableExtrinsic } from '@kiltprotocol/types'
+import { u64 } from '@polkadot/types'
+import { AnyNumber } from '@polkadot/types/types'
+import { KeyId } from './types.chain'
+
+export type KeypairType = string
+
+export interface IPublicKey {
+  publicKey: Uint8Array
+  type: KeypairType
+  // do we need the identifier?
+}
+
+export type IKeyPair = IPublicKey
+
+export type ISigningKey = IPublicKey
+export type IEncryptionKey = IPublicKey
+
+export interface ISigningKeyPair extends ISigningKey {
+  sign: (message: string | Uint8Array) => Uint8Array
+}
+
+export interface IEncryptionKeyPair extends IEncryptionKey {
+  decrypt: (
+    message: Uint8Array,
+    additionalData: Record<string, unknown>
+  ) => Uint8Array | null
+}
+
+// TODO: should this use the same keys as the chain type?
+export interface KeySet {
+  authentication: ISigningKeyPair
+  encryption: IEncryptionKeyPair
+  attestation?: ISigningKeyPair
+  delegation?: ISigningKeyPair
+}
+
+export interface PublicKeySet {
+  authentication: ISigningKey
+  encryption: IEncryptionKey
+  attestation?: ISigningKey
+  delegation?: ISigningKey
+}
+
+export interface TypedPublicKey {
+  type: KeypairType
+  publicKeyHex: string
+}
+
+export interface KeyDetails extends TypedPublicKey {
+  id: string
+  includedAt: number
+}
+
+export interface IDidRecord {
+  did: IIdentity['address']
+  authenticationKey: KeyDetails
+  keyAgreementKeys: KeyDetails[]
```

Comment:
Can this syntax also refer to sets? Or does it only represent an array? In the former case, we might want to change it with a set?

Reply:
Simply because we might offer O(1) retrieval of a specific key, instead of always iterating over the array looking for that very one key.

Reply:
Sets in js are fairly useless for this purpose I think. There is no indexing for sets (but there is for arrays), sets are not ordered - all they offer is certainty that each value is only in there once
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set